docs: explain patch_inner_model and patch_causal_lm_model more (#1133)
NVIDIA-NeMo/Automodel: docs/guides/pipelining.md

Reviewer: ```suggestion
This guide walks you through the complete process of using AutoPipeline for Hugging Face models and the functional API for custom models. You'll learn how to configure pipeline stages, integrate with existing training workflows, optimize performance, and combine pipeline parallelism with other parallelization strategies.
```

Reviewer: ```suggestion
For a complete guide and additional options, please consult the Automodel [Installation Guide](./installation.md).
```

Reviewer: ```suggestion
- **Universal Hugging Face Support**: Works with any Hugging Face decoder-only causal language model including Llama, Qwen, Mistral, Gemma, and more
```

Reviewer: ```suggestion
AutoPipeline is NeMo Automodel's high-level pipeline parallelism interface specifically designed for Hugging Face models, making pipeline parallelism as simple as data parallelism. Built on PyTorch's native `torch.distributed.pipelining`, AutoPipeline provides seamless pipeline parallelism support for any Hugging Face decoder-only causal language model with minimal code changes.
```

Reviewer: ```suggestion
This guide walks you through the complete process of using AutoPipeline for Hugging Face models and the functional API for custom models. You'll learn how to configure pipeline stages, integrate with existing training workflows, optimize performance, and combine pipeline parallelism with other parallelization strategies.
```

Reviewer: ```suggestion
## Prerequisites

Before proceeding with this guide, please ensure that you have NeMo Automodel installed on your machine.

```bash
# Install uv from https://docs.astral.sh/uv/getting-started/installation/
# Initialize the virtual environment using uv
uv venv

# Install the latest stable release from PyPI
uv pip install nemo-automodel

# Or install from source for the latest features
uv pip install git+https://github.com/NVIDIA-NeMo/Automodel.git
```

For a complete guide and additional options please consult the Automodel [Installation Guide](./installation.md).
```

Reviewer: ```suggestion
- **Universal Hugging Face Support**: Works with any Hugging Face decoder-only causal language model including Llama, Qwen, Mistral, Gemma, and more
```

Reviewer: ```suggestion
Here's a minimal example to get started with AutoPipeline using 2 pipeline stages with a Hugging Face model:
```

Reviewer: ```suggestion
While AutoPipeline is specifically designed as a high-level interface for Hugging Face models, the functional API in `nemo_automodel.components.distributed.pipelining.functional` provides more modular and accessible building blocks that can be used with any PyTorch model, including custom architectures. This separation allows for cleaner code organization where AutoPipeline handles Hugging Face-specific optimizations while the functional module remains model-agnostic.
```

Reviewer: ```suggestion
#### Full Hybrid: PP + DP + TP (8 GPUs Total)
```

Reviewer: ```suggestion
AutoPipeline and the functional API together provide a complete pipeline parallelism solution for both Hugging Face and custom models. AutoPipeline offers a high-level, optimized interface specifically for Hugging Face models, while the functional module provides modular, accessible building blocks for custom architectures.
```

Reviewer: ```suggestion
- AutoPipeline provides a simple API for Hugging Face models with powerful customization options
```

Reviewer: ```suggestion
## Prerequisites

Before proceeding with this guide, please ensure that you have NeMo Automodel installed on your machine.

```bash
# Install uv from https://docs.astral.sh/uv/getting-started/installation/
# Initialize the virtual environment using uv
uv venv

# Install the latest stable release from PyPI
uv pip install nemo-automodel

# Or install from source for the latest features
uv pip install git+https://github.com/NVIDIA-NeMo/Automodel.git
```
```

Reviewer: ```suggestion
:::{important}
Before proceeding with this guide, please ensure that you have NeMo AutoModel installed on your machine.
:::

```

# Or install from source for the latest features
uv pip install git+https://github.com/NVIDIA-NeMo/Automodel.git
## Prerequisites

Before proceeding with this guide, please ensure that you have NeMo Automodel installed on your machine.

```bash
# Install uv from https://docs.astral.sh/uv/getting-started/installation/
# Initialize the virtual environment using uv
uv venv

# Install the latest stable release from PyPI
uv pip install nemo-automodel

# Or install from source for the latest features
uv pip install git+https://github.com/NVIDIA-NeMo/Automodel.git

For a complete guide and additional options please consult the AutoModel [Installation Guide](./installation.md).

> For a complete guide and additional options please consult the AutoModel [Installation Guide](./installation.md).

Reviewer: Line 45/46 are duplicates. Pls delete 46.

uv pip install git+https://github.com/NVIDIA-NeMo/Automodel.git
```

> Before proceeding with this guide, please ensure that you have NeMo Automodel installed on your machine.

Reviewer: ```suggestion
Before proceeding with this guide, please ensure that you have NeMo AutoModel installed on your machine.
```

    layers_per_stage=None,          # Layers per stage (None for auto)
    module_fqns_per_model_part=None,  # Manual module assignment

    # Model patching (HF-specific)
    patch_inner_model=True,         # Make decoder forward stage-friendly
    patch_causal_lm_model=True,     # Make CausalLM wrapper return tensors (hidden/logits)
).build(model, loss_fn=loss_fn)
```

### Model patching (`patch_inner_model`, `patch_causal_lm_model`)

> AutoPipeline splits a model by deep-copying it per stage and pruning away modules that don't belong to that stage. Many HuggingFace models assume the full module tree is present and return `ModelOutput` objects; after pruning, their original `forward()` often breaks (or returns objects that are awkward to pipeline).

Reviewer: ```suggestion
AutoPipeline splits a model by deep-copying it per stage and pruning away modules that don't belong to that stage. Many Hugging Face models assume the full module tree is present and return `ModelOutput` objects; after pruning, their original `forward()` often breaks (or returns objects that are awkward to pipeline).
```

> AutoPipeline splits a model by deep-copying it per stage and pruning away modules that don't belong to that stage. Many HuggingFace models assume the full module tree is present and return `ModelOutput` objects; after pruning, their original `forward()` often breaks (or returns objects that are awkward to pipeline).

Reviewer: ```suggestion
AutoPipeline splits a model by deep-copying it per stage and pruning away modules that don't belong to that stage. Many Hugging Face models assume the full module tree is present and return `ModelOutput` objects; after pruning, their original `forward()` often breaks (or returns objects that are awkward to pipeline).
```

  - **Stage 0** (has `embed_tokens`): takes token IDs and produces hidden states.
  - **Middle stages** (no `embed_tokens`): take hidden states from the previous stage (via `inputs_embeds`, or a float tensor passed through `input_ids`) and produce hidden states.
  - Handles sliced layer containers (e.g., `layers` becoming dict-like after stage pruning) and returns a **tensor** of hidden states so stages can be chained.

  For compilation/performance, this patched forward prefers a precomputed `causal_mask_mapping` dict (it will fall back to computing masks and warn if you don't provide it).

- **`patch_causal_lm_model`**: patches the *`...ForCausalLM` wrapper* forward (the module that owns `lm_head`) so pipeline stages return tensors:
  - Returns **hidden states** when `lm_head` is absent on that stage.
  - Returns **logits** when `lm_head` is present (typically only the last stage).
  - Supports `logits_to_keep` to compute logits for only the last `k` tokens.
  
  Note: this is only used when the module you pipeline is a `...ForCausalLM`-style wrapper (i.e., it has a `.model` attribute). If you pass a base decoder module directly, `patch_causal_lm_model` typically has no effect.

> #### When should I change these?

Reviewer: ```suggestion
#### When Should I Change These?
```

> - **Leave both `True` (default)** for standard HuggingFace `AutoModelForCausalLM` / `...ForCausalLM` models. This is the common case and gives the expected behavior: token IDs -> hidden states -> logits across stages.

Reviewer: ```suggestion
- **Leave both `True` (default)** for standard Hugging Face `AutoModelForCausalLM` / `...ForCausalLM` models. This is the common case and gives the expected behavior: token IDs -> hidden states -> logits across stages.
```

- **Set both `False`** when your model already has a pipeline-friendly forward (returns tensors and can accept hidden states when embeddings are absent) or it needs custom kwargs/paths that the HF patch doesn't preserve (common for NeMo AutoModel-native model implementations, packed-sequence/`thd` paths, extra args like `padding_mask`, etc.). Many benchmark configs for NeMo-native models do this (for example `examples/benchmark/configs/qwen3_moe_30b_torch.yaml`).
- **Set `patch_inner_model=False, patch_causal_lm_model=True`** when your inner model is already stage-friendly, but the wrapper forward still returns a `ModelOutput` and you only want the wrapper simplified to “hidden states or logits”.

If you disable `patch_causal_lm_model`, your last stage will typically output hidden states instead of logits; in that case, make sure your `loss_fn` (or your last-stage module) applies the LM head explicitly.

> ### Automatic vs Manual Layer Distribution

Reviewer: ```suggestion
### Automatic vs. Manual Layer Distribution
```
